src/world/storage.rs

```diff
@@ -1,80 +1,31 @@
-use crate::types::bit;
-use crate::types::hash::FNVHash;
 use crate::world::block;
 
-use std::collections::HashMap;
-use std::hash::BuildHasherDefault;
-
 #[derive(Clone)]
 pub struct BlockStorage {
-    blocks: bit::Map,
-    block_map: Vec<(block::Block, u32)>,
-    rev_block_map: HashMap<block::Block, usize, BuildHasherDefault<FNVHash>>,
+    blocks: Vec<block::Block>,
 }
 
 impl BlockStorage {
-    pub fn new(size: usize) -> BlockStorage {
+    pub fn new(size: usize) -> Self {
         Self::new_default(size, block::Air {})
     }
 
-    pub fn new_default(size: usize, def: block::Block) -> BlockStorage {
-        let mut storage = BlockStorage {
-            blocks: bit::Map::new(size, 4),
-            block_map: vec![(def, size as u32)],
-            rev_block_map: HashMap::with_hasher(BuildHasherDefault::default()),
-        };
-        storage.rev_block_map.insert(def, 0);
-        storage
+    pub fn new_default(size: usize, def: block::Block) -> Self {
+        Self {
+            blocks: vec![def; size],
+        }
     }
 
     pub fn get(&self, idx: usize) -> block::Block {
-        let idx = self.blocks.get(idx);
-        self.block_map[idx].0
+        self.blocks[idx]
     }
 
     pub fn set(&mut self, idx: usize, b: block::Block) -> bool {
-        use std::collections::hash_map::Entry;
-        let old = self.get(idx);
-        if old == b {
-            return false;
-        }
-        // Clean up the old block
-        {
-            let idx = *self.rev_block_map.get(&old).unwrap();
-            let info = &mut self.block_map[idx];
-            info.1 -= 1;
-            if info.1 == 0 {
-                // None left of this type
-                self.rev_block_map.remove(&old);
-            }
-        }
-
-        if let Entry::Vacant(entry) = self.rev_block_map.entry(b) {
-            let mut found = false;
-            let id = entry.insert(self.block_map.len());
-            for (i, ref mut info) in self.block_map.iter_mut().enumerate() {
-                if info.1 == 0 {
-                    info.0 = b;
-                    *id = i;
-                    found = true;
-                    break;
-                }
-            }
-            if !found {
-                if self.block_map.len() >= 1 << self.blocks.bit_size {
-                    let new_size = self.blocks.bit_size + 1;
-                    self.blocks = self.blocks.resize(new_size);
-                }
-                self.block_map.push((b, 0));
-            }
-        }
-
-        {
-            let b_idx = self.rev_block_map[&b];
-            let info = &mut self.block_map[b_idx];
-            info.1 += 1;
-            self.blocks.set(idx, b_idx);
+        if self.blocks[idx] == b {
```

Comment:
this reminds me if CAS any chance we could make this atomic, do you know if Block is of size <= 8 bytes?

Reply:
It's currently 10 bytes based on (on my x86_64 system at least). I haven't done too much with atomics in rust before. I wouldn't have thought it would be required here. Is it for performance benefits?

The other option is always setting the block without checking. The only downside is that it would always set the dirty flag, but it seems unlikely for a block to be replaced with the same block, outside of loading chunks.

Reply:
Doing it conditionless seems pretty tempting because having one less branch in what i assume to be hot code would be pretty nice for such minor possible overhead, also the CAS i mentioned is probably irrelevant anyways as long as we wouldn't profit from a BlockStorage with interior mutability and as long as the Vec has to be able to grow dynamically.